core/modules/map.rs

```diff
@@ -55,46 +61,79 @@ pub(crate) enum SymbolicModule {
-  // Handling of specifiers and v8 objects
-  pub handles: Vec<v8::Global<v8::Module>>,
-  pub info: Vec<ModuleInfo>,
-  pub(crate) by_name_js: HashMap<ModuleName, SymbolicModule>,
-  pub(crate) by_name_json: HashMap<ModuleName, SymbolicModule>,
-  pub(crate) next_load_id: ModuleLoadId,
-
   // Handling of futures for loading module sources
-  pub loader: Rc<dyn ModuleLoader>,
-  pub(crate) dynamic_import_map:
-    HashMap<ModuleLoadId, v8::Global<v8::PromiseResolver>>,
-  pub(crate) preparing_dynamic_imports:
-    FuturesUnordered<Pin<Box<PrepareLoadFuture>>>,
-  pub(crate) pending_dynamic_imports:
-    FuturesUnordered<StreamFuture<RecursiveModuleLoad>>,
-
-  // This store is used temporarily, to forward parsed JSON
-  // value from `new_json_module` to `json_module_evaluation_steps`
+  // TODO(mmastrac): we should not be swapping this loader out
+  pub(crate) loader: RefCell<Rc<dyn ModuleLoader>>,
+
+  dynamic_import_map:
+    RefCell<HashMap<ModuleLoadId, v8::Global<v8::PromiseResolver>>>,
+  preparing_dynamic_imports:
+    RefCell<FuturesUnordered<Pin<Box<PrepareLoadFuture>>>>,
+  pending_dynamic_imports:
+    RefCell<FuturesUnordered<StreamFuture<RecursiveModuleLoad>>>,
+  pending_dyn_mod_evaluate: RefCell<Vec<DynImportModEvaluate>>,
+  data: RefCell<ModuleMapData>,
+
+  /// A counter used to delay our dynamic import deadlock detection by one spin
+  /// of the event loop.
+  pub(crate) dyn_module_evaluate_idle_counter: Cell<u32>,
+}
+
+#[derive(Default)]
+pub(crate) struct ModuleMapData {
+  /// Inverted index from module to index in `info`.
+  handles_inverted: HashMap<v8::Global<v8::Module>, usize>,
+  /// The handles we have loaded so far, corresponding with the [`ModuleInfo`] in `info`.
+  handles: Vec<v8::Global<v8::Module>>,
+  /// The modules we have loaded so far.
+  info: Vec<ModuleInfo>,
+  /// [`ModuleName`] to [`SymbolicModule`] for JS/WASM modules.
+  by_name_js: HashMap<ModuleName, SymbolicModule>,
+  /// [`ModuleName`] to [`SymbolicModule`] for JSON modules.
+  by_name_json: HashMap<ModuleName, SymbolicModule>,
+  /// The next ID used for a load.
+  next_load_id: ModuleLoadId,
+  /// If a main module has been loaded, points to it by index.
+  main_module_id: Option<ModuleId>,
+  /// This store is used temporarily, to forward parsed JSON
+  /// value from `new_json_module` to `json_module_evaluation_steps`
   json_value_store: HashMap<v8::Global<v8::Module>, v8::Global<v8::Value>>,
 }
 
 impl ModuleMap {
-  pub fn collect_modules(
+  pub(crate) fn next_load_id(&self) -> i32 {
+    // TODO(mmastrac): move recursive module loading into here so we can avoid making this pub
+    let mut data = self.data.borrow_mut();
+    let id = data.next_load_id;
+    data.next_load_id += 1;
+    id + 1
+  }
+
+  fn collect_modules(
     &self,
-  ) -> Vec<(AssertedModuleType, &ModuleName, &SymbolicModule)> {
-    let mut output = vec![];
+    mut f: impl FnMut(usize, AssertedModuleType, &ModuleName, &SymbolicModule),
```

Comment:
Can you add a comment about this param? It's not obvious what it's meant to do. Maybe this method should be renamed from `collect_modules()` if it doesn't return anything?